Use PCC bounds from .caprelocs / .rela.dyn

bsdjhb opened the pull request:
Note that for CHERI-RISC-V, this depends on changes in the c18n toolchain that are not yet in the packaged version.  My assumption is that for our next release we will be requiring a newer c18n toolchain anyway, so we can assume that if `PT_CHERI_PCC` is present, we are using an updated toolchain that writes caprelocs with the per-compartment PCC bounds.

We probably need to get an updated c18n package built (after my rebasing work finishes) before landing the last change.

bsdjhb committed:
<machine/cherireg.h>: Define CHERI_RODATA_PTR helper macro

This macro is responsible for returning a valid capability before any
relocations have been performed.  It is intended for use in early
relocation code.

bsdjhb committed:
csu: Use CHERI_RODATA_PTR in place of RODATA_PTR

bsdjhb committed:
sys: Refactor initial cap relocs handling for the kernel

Reuse the init_linker_file_cap_relocs function to process cap relocs
for the kernel from locore instead of using cheri_init_globals_3.
This will permit customizing the behavior in future changes.

Add a new callback for init_linker_file_cap_relocs for this case that
accepts a single RWX capability as its argument which locore derives
from PCC.

For RISC-V, set the flags to 0 in this capability to avoid leaking
them into data capabilities.  This does require setting cap mode
explicitly for function capabilities in the callback.

bsdjhb committed:
sys: Adjust support for bounded code pointers

Prepare for preferring the bounds from cap relocs and relative ELF
relocations when PT_CHERI_PCC is present.  For now the bounds are
always disabled, but this provides the hooks to enable this in the
future when support for PT_CHERI_PCC is added to the kernel linker.

bsdjhb committed:
rtld: Use tight PCC bounds if PT_CHERI_PCC is present

This trusts the bounds from both cap relocs and relative ELF
relocations if PT_CHERI_PCC bounds program headers are present.

bsdjhb committed:
csu: Use tight PCC bounds for static binaries when PT_CHERI_PCC is present

For caprelocs, use cheri_init_globals_impl() instead of
cheri_init_globals_3().  This requires using CHERI_RODATA_PTR to
lookup the start and and end of the caprelocs section.

For ELF relocations, trust the bounds from relative relocations for
code pointers if PT_CHERI_PCC is present.

bsdjhb commented:
Since the bounds cannot be used on Morello without PT_CHERI_PCC due to a bug in Morello lld (it's PCC bounds didn't cover .data.rel.ro), I've reworked this to always require PT_CHERI_PCC.  It now also doesn't try to do arch-specific commits, but instead static binaries (csu) vs userspace runtime loader vs kernel.  The kernel bits are incomplete as I don't want to duplicate Konrad's work.  However, once Konrad's changes are rebased, we can perhaps pull out the changes to teach the kernel linker about PT_CHERI_PCC into `dev` separately from some of the other kernel c18n changes.

jrtc27 commented on sys/cheri/cheri_cap_relocs.c:
```diff
@@ -39,7 +40,65 @@ extern void init_cap_relocs(void *data_cap, void *code_cap);
+			    CHERI_PERMS_KERNEL_CODE);
+#ifdef CHERI_FLAGS_CAP_MODE
+			src = __builtin_cheri_flags_set(src,
+			    CHERI_FLAGS_CAP_MODE);
```
This shouldn't be done for hybrid, but also the expectation is that code_cap already has the right flags (cheri_init_globals_3 doesn't set the flags).

bsdjhb replied:
Note that this code is only used from purecap and not from hybrid (I guess we don't have any caprelocs in hybrid?)

bsdjhb commented on sys/cheri/cheri_cap_relocs.c:
```diff
@@ -39,7 +40,65 @@ extern void init_cap_relocs(void *data_cap, void *code_cap);
 init_cap_relocs(void *data_cap, void *code_cap)
 {
-	cheri_init_globals_3(data_cap, code_cap, data_cap);
+	const struct capreloc *start_relocs, *stop_relocs;
```
I ended up adding a callback for the kernel and using `cllc` to create the callback pointer.  That does seem to work.  I will upload it as a fixup commit in a bit so you can decide if you prefer it.

bsdjhb committed:
csu: Declare __rela_dyn_* as arrays

This requires fixing RODATA_PTR to decay arrays to pointers.

Co-authored-by: Jessica Clarke <jrtc27@jrtc27.com>

bsdjhb committed:
kld: Add ef_symbol_address helper routine

bsdjhb committed:
kld: Use PCC bounds from PT_CHERI_PCC segments when they exist

- Add an array of PCC capabilities to elf_file_t that are used to
  derive capabilities similar to mapbase.  If PT_CHERI_PCC segments
  are present, use the matching PCC capability as the base capability
  for function symbols.

- If PT_CHERI_PCC is present, set a new LINKER_FILE_PCC_BOUNDS flag
  on the corresponding linker_file and use this flag to use bounds
  for function relative relocations.  I initially tried using an
  internal flag in elf_file_t and a new linker_if.m method, but
  kobj methods cannot be used from elf_reloc since it can be called
  to resolve ifuncs before kobj is initialized.

- Walk the program headers to check for PT_CHERI_PCC in elf_reloc_self
  on Morello.  This assumes that the ELF header is mapped at the start
  of the kernel at KERNBASE.

- PT_CHERI_PCC segments do not currently work on RISC-V kernels as
  RISC-V kernels do not map program headers into memory due to SBI
  direct-boot.

jrtc27 commented on lib/csu/aarch64c/crt1_c.c:
```diff
@@ -48,12 +48,12 @@
-	    : "=C" (_p));						\
+	    "adrp %0, %c1\n\t"						\
+	    "add %0, %0, :lo12:%c1\n\t"					\
+		: "=C" (_p) : "i" (x));					\
```
The indentation was fine, no?

jrtc27 commented on sys/arm64/include/cherireg.h:
```diff
@@ -222,4 +222,21 @@
+		: "=C" (_p) : "i" (x));					\
+	_p; })
+#else
+#define	CHERI_RODATA_PTR(x)	(x)
```
This doesn't decay or do &* like the rtld version

jrtc27 reviewed (commented):
The commit message for "sys: Inline cap relocs handling for the kernel" seems outdated, the callback's already there, it's just switching to using it for self-relocation.

Also this needs a sweep to switch to the new cheri_noun_verb macro names (the old names are still available for now...).

bsdjhb commented:
> The commit message for "sys: Inline cap relocs handling for the kernel" seems outdated, the callback's already there, it's just switching to using it for self-relocation.

Ok, I've reworded it.

> Also this needs a sweep to switch to the new cheri_noun_verb macro names (the old names are still available for now...).

Hmm, I thought I had mostly done this when fixing the conflicts while rebasing across that change, but it looks like I mismerged at least one in rtld for aarch64.

jrtc27 commented on sys/riscv/riscv/locore.S:
```diff
@@ -396,11 +396,19 @@ va:
+	csetflags ca5, ca4, zero
+	cllc	cra, _C_LABEL(init_linker_file_cap_relocs)
 	cjalr	cra
+	cincoffset csp, csp, (2 * 16)
```
Did an earlier version spill to the stack or something?

bsdjhb replied:
Oh, humm, yes.  Originally I had to create a structure with two capabilities to pass as the `arg` to the callback, but I later reworked it to only pass in a single capability directly as `arg`.

jrtc27 commented on sys/arm64/arm64/elf_machdep.c:
```diff
@@ -761,10 +767,24 @@ arm64_exec_protect(struct image_params *imgp, int flags __unused)
 	size_t rela_size = 0;
+	bool use_code_bounds = false;
+
+	/* Assume ELF header is at KERNBASE. */
```
Should we check hdr points to ELF magic before interpreting its fields and looking for program headers?

jrtc27 commented on sys/kern/link_elf.c:
```diff
@@ -433,15 +434,57 @@ static caddr_t
+	}
+#endif
+	pccbase = ef->mapbase;
+	hdr = (const Elf_Ehdr *)ef->mapbase;
```
Could do a general "if kernel file and hdr does not point to ELF magic then return" rather than hard-coding it as a RISC-V thing above

bsdjhb replied:
I just checked for an ELF header always.

jrtc27 commented on sys/kern/link_elf.c:
```diff
@@ -1329,6 +1392,13 @@ link_elf_load_file(linker_class_t cls, const char* filename,
+		ef->pccbase = cheri_bounds_set(ef->pccbase, phpcc->p_memsz);
+	} else {
+		ef->pccbase = mapbase;
+	}
```
```c
	ef->pccbase = mapbase;
	if (phpcc != NULL) {
		lf->flags |= LINKER_FILE_PCC_BOUNDS;
		ef->pccbase += phpcc->p_vaddr - base_vaddr;
		ef->pccbase = cheri_bounds_set(ef->pccbase, phpcc->p_memsz);
	}
```
might be a bit nicer?

bsdjhb commented:
I went ahead and added a followup change to handle multiple PT_CHERI_PCC segments modeled on RTLD.  If that looks ok, I can just squash it into the last commit.

@kwitaszczyk this is a bit different than the version in your branch, but I think it's a bit cleaner.  It doesn't have a static limit on the number of segments unlike the kernel_c18n version.

jrtc27 commented on sys/kern/link_elf.c:
```diff
@@ -83,7 +83,8 @@ typedef struct elf_file {
 	caddr_t		mapbase;	/* Capability for derived pointers */
-	caddr_t		pccbase;	/* Capability for code pointers */
+	caddr_t		*pcc_caps;	/* Capability for code pointers */
+	unsigned long	npcc_caps;
```
size_t here and for i/j below? Upstream may be bad about using meaningful types but we don't have to make that worse :)

bsdjhb replied:
Hmm, I can.  I used `unsigned long` in rtld already and was matching that, so I would probably need to fix rtld to match.

jrtc27 replied:
Eh, ok, let's just leave it then

jrtc27 reviewed (approved):
(once squashed)
